### PR TITLE
full ref support

### PR DESCRIPTION
**Pull request by dzmitry-lahoda**
I hope. Please look into reasons I have added ref parameter index

### CODE REVIEW

File: src/FastExpressionCompiler/FastExpressionCompiler.cs [hunk 1]
```diff
@@ -2549,6 +2554,31 @@ private static bool TryEmitMethodCall(object exprObj, object[] paramExprs, ILGen
...
+                IList<Expression> refed = null;
+                var receivingParameters = expr.Method.GetParameters();
+                for (int i = 0; i < expr.Method.GetParameters().Length; i++)
+                {
+                    if (receivingParameters[i].ParameterType.IsByRef)
+                    {
+                        if (refed == null)
+                            refed = new List<Expression>(expr.Arguments);
+                        var passed = expr.Arguments[i] as ParameterExpression;
+                        if (!passed.IsByRef)
```
Comment (dadhi) [inline src/FastExpressionCompiler/FastExpressionCompiler.cs]: `passed` should be check for `null` via `?.`.
the original argument should be returned.

Then the result may be changed to `IEnimerable<object>`. Is it possible?

File: src/FastExpressionCompiler/FastExpressionCompiler.cs [hunk 2]
```diff
@@ -2549,6 +2554,31 @@ private static bool TryEmitMethodCall(object exprObj, object[] paramExprs, ILGen
...
+
+            // if call is done into byref method parameters there is no indicators in tree, so grab that from method
+            // current approach is to copy into new list only if there are by ref with by ref parameters,
+            // possible approach to store hit map of small size (possible 256 bit #89) to check if parameter is by ref
+            // https://stackoverflow.com/questions/12658883/what-is-the-maximum-number-of-parameters-that-a-c-sharp-method-can-be-defined-as
+            private static IList<Expression> MakeByRefParameters(MethodCallExpression expr)
+            {
+                IList<Expression> refed = null;
+                var receivingParameters = expr.Method.GetParameters();
+                for (int i = 0; i < expr.Method.GetParameters().Length; i++)
```
Comment (dadhi) [inline src/FastExpressionCompiler/FastExpressionCompiler.cs]: `expr.Method.GetParameters().Length` will be evaluated on each iteration, move it to var before loop.

**Comment by dadhi** [comment]
Btw @dzmitry-lahoda, good job as always. 

I have looked back to related code from previous merge, and found this repetition of `IsByRef..` call:

```
    if (Tools.IsByRefParameter(left))
        EmitLoadParamArg(il, paramIndex, false);

    if (!TryEmit(right, rightNodeType, exprType, paramExprs, il, ref closure, ExpressionType.Assign))
         return false;

    if (Tools.IsByRefParameter(left))
```

**Comment by dadhi** [comment]
In `GetParamExprTypes` there is a multiple access to param types by index in single statement, it should be refactored to variable.

**Comment by dadhi** [comment]
In `ParameterExpressionInfo` conversion to expression does not consider `IsByRef` field, probably it should.

**Comment by dadhi** [comment]
I have a strong feeling that `IsByReferenceReturn` method itself should adjusted to fully support ExprInfo for Block and Assign.

Both switch cases inside loop are boiling down to extracting Assign Left operand, so both cases may be consolidated further.

**Comment by dadhi** [comment]
Oops, and `exprObj.ToExpression();` is done in a loop in `IsByReferenceReturn` method. Should be moved outside.

**Comment by dzmitry-lahoda** [comment]
>> In ParameterExpressionInfo conversion to expression does not consider IsByRef field, probably it should.
I see that `ParameterExpression` has inheritance and switch by type implementation in .NET. Not sure how to properly fit it into FEC. I guess `IsByRef` would work already, also storing IsByRef  in field may be not the best. I do not know what is propert to do, but probably stuff works. I guess there should be more guidance-benchmarks of `Info` stuff implementation.

```
public static ParameterExpression Parameter(Type type, string name)
{
	ContractUtils.RequiresNotNull(type, "type");
	if (type == typeof(void))
	{
		throw Error.ArgumentCannotBeOfTypeVoid();
	}
	bool isByRef = type.IsByRef;
	if (isByRef)
	{
		type = type.GetElementType();
	}
	return ParameterExpression.Make(type, name, isByRef);
}

internal sealed class ByRefParameterExpression : TypedParameterExpression
{
	internal ByRefParameterExpression(Type type, string name)
		: base(type, name)
	{
	}

	internal override bool GetIsByRef()
	{
		return true;
	}
}

```

**Comment by dadhi** [comment]
Ok, regarding ParamExprInfo I wull look deepe. If byref is derived from parameter type automatically, then fine.

**Comment by dzmitry-lahoda** [comment]
> Oops, and exprObj.ToExpression(); is done in a loop in IsByReferenceReturn method. Should be moved outside.

This was `Any()`, but with bug. Fixed.

**Comment by dadhi** [comment]
Dima, any other updates on this PR or I can merge?

**Comment by dzmitry-lahoda** [comment]
Please merge.

**Pull request merged**